src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Measurement.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Measurement stores one observed metrics value and its associated tags. This type is used by Observable instruments' Observe() method when reporting current measurements.
+    /// with the associated tags.
+    /// </summary>
+#if ALLOW_PARTIALLY_TRUSTED_CALLERS
+        [System.Security.SecuritySafeCriticalAttribute]
+#endif
+    public readonly struct Measurement<T> where T : struct
+    {
+        private readonly KeyValuePair<string, object?>[] _tags;
+
+        /// <summary>
+        /// Initializes a new instance of the Measurement using the value and the list of tags.
+        /// </summary>
+        /// <param name="value">The measurement value.</param>
+        public Measurement(T value)
+        {
+            _tags = Instrument.EmptyTags;
+            Value = value;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the Measurement using the value and the list of tags.
+        /// </summary>
+        /// <param name="value">The measurement value.</param>
+        /// <param name="tags">The measurement associated tags list.</param>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object?>>? tags)
+        {
+            _tags = tags?.ToArray() ?? Instrument.EmptyTags;
```

Comment:
Looks like this is the only Linq usage, seems reasonable given other overloads.  I guess you always want a copy of the `tags` array?  Any interest in perhaps providing static factory method for folks who wish to avoid that allocation (by allowing shared ownership).

Reply:
This is a not hot path for perf concern. This is why it is ok to allocate every time it gets called. we can try optimizing later if we find any concern which is unlikely to happen.